###################################################################################################################################
###################################################################################################################################
Load to stage in progress
###################################################################################################################################
###################################################################################################################################
20/02/14 17:58:11 WARN Utils: Your hostname, DESKTOP-502C4I4 resolves to a loopback address: 127.0.1.1; using 192.168.1.25 instead (on interface wifi0)
20/02/14 17:58:11 WARN Utils: Set SPARK_LOCAL_IP if you need to bind to another address
20/02/14 17:58:13 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
Using Spark's default log4j profile: org/apache/spark/log4j-defaults.properties
20/02/14 17:58:19 INFO SparkContext: Running Spark version 2.4.5
20/02/14 17:58:20 INFO SparkContext: Submitted application: MySession
20/02/14 17:58:20 INFO SecurityManager: Changing view acls to: devetluser
20/02/14 17:58:20 INFO SecurityManager: Changing modify acls to: devetluser
20/02/14 17:58:20 INFO SecurityManager: Changing view acls groups to: 
20/02/14 17:58:20 INFO SecurityManager: Changing modify acls groups to: 
20/02/14 17:58:20 INFO SecurityManager: SecurityManager: authentication disabled; ui acls disabled; users  with view permissions: Set(devetluser); groups with view permissions: Set(); users  with modify permissions: Set(devetluser); groups with modify permissions: Set()
20/02/14 17:58:21 INFO Utils: Successfully started service 'sparkDriver' on port 3086.
20/02/14 17:58:21 INFO SparkEnv: Registering MapOutputTracker
20/02/14 17:58:21 INFO SparkEnv: Registering BlockManagerMaster
20/02/14 17:58:21 INFO BlockManagerMasterEndpoint: Using org.apache.spark.storage.DefaultTopologyMapper for getting topology information
20/02/14 17:58:21 INFO BlockManagerMasterEndpoint: BlockManagerMasterEndpoint up
20/02/14 17:58:21 INFO DiskBlockManager: Created local directory at /tmp/blockmgr-1ee836c7-e4ed-4b26-9636-b3026605f2d6
20/02/14 17:58:22 INFO MemoryStore: MemoryStore started with capacity 366.3 MB
20/02/14 17:58:22 INFO SparkEnv: Registering OutputCommitCoordinator
20/02/14 17:58:22 INFO Utils: Successfully started service 'SparkUI' on port 4040.
20/02/14 17:58:22 INFO SparkUI: Bound SparkUI to 0.0.0.0, and started at http://192.168.1.25:4040
20/02/14 17:58:22 INFO Executor: Starting executor ID driver on host localhost
20/02/14 17:58:23 INFO Utils: Successfully started service 'org.apache.spark.network.netty.NettyBlockTransferService' on port 3091.
20/02/14 17:58:23 INFO NettyBlockTransferService: Server created on 192.168.1.25:3091
20/02/14 17:58:23 INFO BlockManager: Using org.apache.spark.storage.RandomBlockReplicationPolicy for block replication policy
20/02/14 17:58:23 INFO BlockManagerMaster: Registering BlockManager BlockManagerId(driver, 192.168.1.25, 3091, None)
20/02/14 17:58:23 INFO BlockManagerMasterEndpoint: Registering block manager 192.168.1.25:3091 with 366.3 MB RAM, BlockManagerId(driver, 192.168.1.25, 3091, None)
20/02/14 17:58:23 INFO BlockManagerMaster: Registered BlockManager BlockManagerId(driver, 192.168.1.25, 3091, None)
20/02/14 17:58:23 INFO BlockManager: Initialized BlockManager: BlockManagerId(driver, 192.168.1.25, 3091, None)
20/02/14 17:58:24 INFO SharedState: Setting hive.metastore.warehouse.dir ('null') to the value of spark.sql.warehouse.dir ('file:/edw/Reviews/SourceCode/bash/spark-warehouse').
20/02/14 17:58:24 INFO SharedState: Warehouse path is 'file:/edw/Reviews/SourceCode/bash/spark-warehouse'.
20/02/14 17:58:26 INFO StateStoreCoordinatorRef: Registered StateStoreCoordinator endpoint
20/02/14 17:58:27 INFO InMemoryFileIndex: It took 337 ms to list leaf files for 1 paths.
20/02/14 17:58:27 INFO InMemoryFileIndex: It took 5 ms to list leaf files for 1 paths.
20/02/14 17:58:32 INFO FileSourceStrategy: Pruning directories with: 
20/02/14 17:58:32 INFO FileSourceStrategy: Post-Scan Filters: 
20/02/14 17:58:32 INFO FileSourceStrategy: Output Data Schema: struct<value: string>
20/02/14 17:58:32 INFO FileSourceScanExec: Pushed Filters: 
20/02/14 17:58:34 INFO CodeGenerator: Code generated in 480.6839 ms
20/02/14 17:58:34 INFO MemoryStore: Block broadcast_0 stored as values in memory (estimated size 282.7 KB, free 366.0 MB)
20/02/14 17:58:34 INFO MemoryStore: Block broadcast_0_piece0 stored as bytes in memory (estimated size 23.3 KB, free 366.0 MB)
20/02/14 17:58:34 INFO BlockManagerInfo: Added broadcast_0_piece0 in memory on 192.168.1.25:3091 (size: 23.3 KB, free: 366.3 MB)
20/02/14 17:58:34 INFO SparkContext: Created broadcast 0 from json at NativeMethodAccessorImpl.java:0
20/02/14 17:58:34 INFO FileSourceScanExec: Planning scan with bin packing, max size: 22011312 bytes, open cost is considered as scanning 4194304 bytes.
20/02/14 17:58:35 INFO SparkContext: Starting job: json at NativeMethodAccessorImpl.java:0
20/02/14 17:58:35 INFO DAGScheduler: Got job 0 (json at NativeMethodAccessorImpl.java:0) with 4 output partitions
20/02/14 17:58:35 INFO DAGScheduler: Final stage: ResultStage 0 (json at NativeMethodAccessorImpl.java:0)
20/02/14 17:58:35 INFO DAGScheduler: Parents of final stage: List()
20/02/14 17:58:35 INFO DAGScheduler: Missing parents: List()
20/02/14 17:58:35 INFO DAGScheduler: Submitting ResultStage 0 (MapPartitionsRDD[3] at json at NativeMethodAccessorImpl.java:0), which has no missing parents
20/02/14 17:58:35 INFO MemoryStore: Block broadcast_1 stored as values in memory (estimated size 11.6 KB, free 366.0 MB)
20/02/14 17:58:35 INFO MemoryStore: Block broadcast_1_piece0 stored as bytes in memory (estimated size 6.7 KB, free 366.0 MB)
20/02/14 17:58:35 INFO BlockManagerInfo: Added broadcast_1_piece0 in memory on 192.168.1.25:3091 (size: 6.7 KB, free: 366.3 MB)
20/02/14 17:58:35 INFO SparkContext: Created broadcast 1 from broadcast at DAGScheduler.scala:1163
20/02/14 17:58:35 INFO DAGScheduler: Submitting 4 missing tasks from ResultStage 0 (MapPartitionsRDD[3] at json at NativeMethodAccessorImpl.java:0) (first 15 tasks are for partitions Vector(0, 1, 2, 3))
20/02/14 17:58:35 INFO TaskSchedulerImpl: Adding task set 0.0 with 4 tasks
20/02/14 17:58:35 INFO TaskSetManager: Starting task 0.0 in stage 0.0 (TID 0, localhost, executor driver, partition 0, PROCESS_LOCAL, 8267 bytes)
20/02/14 17:58:35 INFO TaskSetManager: Starting task 1.0 in stage 0.0 (TID 1, localhost, executor driver, partition 1, PROCESS_LOCAL, 8267 bytes)
20/02/14 17:58:35 INFO TaskSetManager: Starting task 2.0 in stage 0.0 (TID 2, localhost, executor driver, partition 2, PROCESS_LOCAL, 8267 bytes)
20/02/14 17:58:35 INFO TaskSetManager: Starting task 3.0 in stage 0.0 (TID 3, localhost, executor driver, partition 3, PROCESS_LOCAL, 8267 bytes)
20/02/14 17:58:35 INFO Executor: Running task 0.0 in stage 0.0 (TID 0)
20/02/14 17:58:35 INFO Executor: Running task 2.0 in stage 0.0 (TID 2)
20/02/14 17:58:35 INFO Executor: Running task 1.0 in stage 0.0 (TID 1)
20/02/14 17:58:35 INFO Executor: Running task 3.0 in stage 0.0 (TID 3)
20/02/14 17:58:35 INFO FileScanRDD: Reading File path: file:///edw/Reviews/SourceCode/data/AmzReviews.json, range: 44022624-66033936, partition values: [empty row]
20/02/14 17:58:35 INFO FileScanRDD: Reading File path: file:///edw/Reviews/SourceCode/data/AmzReviews.json, range: 66033936-83850945, partition values: [empty row]
20/02/14 17:58:35 INFO FileScanRDD: Reading File path: file:///edw/Reviews/SourceCode/data/AmzReviews.json, range: 0-22011312, partition values: [empty row]
20/02/14 17:58:35 INFO FileScanRDD: Reading File path: file:///edw/Reviews/SourceCode/data/AmzReviews.json, range: 22011312-44022624, partition values: [empty row]
20/02/14 17:58:35 INFO CodeGenerator: Code generated in 18.3706 ms
20/02/14 17:58:36 INFO Executor: Finished task 1.0 in stage 0.0 (TID 1). 2307 bytes result sent to driver
20/02/14 17:58:36 INFO Executor: Finished task 3.0 in stage 0.0 (TID 3). 2307 bytes result sent to driver
20/02/14 17:58:36 INFO Executor: Finished task 2.0 in stage 0.0 (TID 2). 2307 bytes result sent to driver
20/02/14 17:58:36 INFO Executor: Finished task 0.0 in stage 0.0 (TID 0). 2307 bytes result sent to driver
20/02/14 17:58:36 INFO TaskSetManager: Finished task 2.0 in stage 0.0 (TID 2) in 1086 ms on localhost (executor driver) (1/4)
20/02/14 17:58:36 INFO TaskSetManager: Finished task 1.0 in stage 0.0 (TID 1) in 1097 ms on localhost (executor driver) (2/4)
20/02/14 17:58:36 INFO TaskSetManager: Finished task 0.0 in stage 0.0 (TID 0) in 1119 ms on localhost (executor driver) (3/4)
20/02/14 17:58:36 INFO TaskSetManager: Finished task 3.0 in stage 0.0 (TID 3) in 1097 ms on localhost (executor driver) (4/4)
20/02/14 17:58:36 INFO TaskSchedulerImpl: Removed TaskSet 0.0, whose tasks have all completed, from pool 
20/02/14 17:58:37 INFO DAGScheduler: ResultStage 0 (json at NativeMethodAccessorImpl.java:0) finished in 1.373 s
20/02/14 17:58:37 INFO DAGScheduler: Job 0 finished: json at NativeMethodAccessorImpl.java:0, took 2.035196 s
20/02/14 17:58:38 INFO FileSourceStrategy: Pruning directories with: 
20/02/14 17:58:38 INFO FileSourceStrategy: Post-Scan Filters: 
20/02/14 17:58:38 INFO FileSourceStrategy: Output Data Schema: struct<asin: string, overall: double, reviewerID: string, reviewerName: string, unixReviewTime: bigint ... 3 more fields>
20/02/14 17:58:38 INFO FileSourceScanExec: Pushed Filters: 
20/02/14 17:58:38 INFO CodeGenerator: Code generated in 59.9943 ms
20/02/14 17:58:38 INFO MemoryStore: Block broadcast_2 stored as values in memory (estimated size 282.7 KB, free 365.7 MB)
20/02/14 17:58:38 INFO MemoryStore: Block broadcast_2_piece0 stored as bytes in memory (estimated size 23.3 KB, free 365.7 MB)
20/02/14 17:58:38 INFO BlockManagerInfo: Added broadcast_2_piece0 in memory on 192.168.1.25:3091 (size: 23.3 KB, free: 366.2 MB)
20/02/14 17:58:38 INFO SparkContext: Created broadcast 2 from jdbc at NativeMethodAccessorImpl.java:0
20/02/14 17:58:38 INFO FileSourceScanExec: Planning scan with bin packing, max size: 22011312 bytes, open cost is considered as scanning 4194304 bytes.
20/02/14 17:58:38 INFO SparkContext: Starting job: jdbc at NativeMethodAccessorImpl.java:0
20/02/14 17:58:38 INFO DAGScheduler: Got job 1 (jdbc at NativeMethodAccessorImpl.java:0) with 4 output partitions
20/02/14 17:58:38 INFO DAGScheduler: Final stage: ResultStage 1 (jdbc at NativeMethodAccessorImpl.java:0)
20/02/14 17:58:38 INFO DAGScheduler: Parents of final stage: List()
20/02/14 17:58:38 INFO DAGScheduler: Missing parents: List()
20/02/14 17:58:38 INFO DAGScheduler: Submitting ResultStage 1 (MapPartitionsRDD[8] at jdbc at NativeMethodAccessorImpl.java:0), which has no missing parents
20/02/14 17:58:38 INFO MemoryStore: Block broadcast_3 stored as values in memory (estimated size 64.3 KB, free 365.6 MB)
20/02/14 17:58:38 INFO MemoryStore: Block broadcast_3_piece0 stored as bytes in memory (estimated size 29.4 KB, free 365.6 MB)
20/02/14 17:58:38 INFO BlockManagerInfo: Added broadcast_3_piece0 in memory on 192.168.1.25:3091 (size: 29.4 KB, free: 366.2 MB)
20/02/14 17:58:38 INFO SparkContext: Created broadcast 3 from broadcast at DAGScheduler.scala:1163
20/02/14 17:58:38 INFO DAGScheduler: Submitting 4 missing tasks from ResultStage 1 (MapPartitionsRDD[8] at jdbc at NativeMethodAccessorImpl.java:0) (first 15 tasks are for partitions Vector(0, 1, 2, 3))
20/02/14 17:58:38 INFO TaskSchedulerImpl: Adding task set 1.0 with 4 tasks
20/02/14 17:58:38 INFO TaskSetManager: Starting task 0.0 in stage 1.0 (TID 4, localhost, executor driver, partition 0, PROCESS_LOCAL, 8267 bytes)
20/02/14 17:58:38 INFO TaskSetManager: Starting task 1.0 in stage 1.0 (TID 5, localhost, executor driver, partition 1, PROCESS_LOCAL, 8267 bytes)
20/02/14 17:58:38 INFO TaskSetManager: Starting task 2.0 in stage 1.0 (TID 6, localhost, executor driver, partition 2, PROCESS_LOCAL, 8267 bytes)
20/02/14 17:58:38 INFO TaskSetManager: Starting task 3.0 in stage 1.0 (TID 7, localhost, executor driver, partition 3, PROCESS_LOCAL, 8267 bytes)
20/02/14 17:58:38 INFO Executor: Running task 2.0 in stage 1.0 (TID 6)
20/02/14 17:58:38 INFO Executor: Running task 1.0 in stage 1.0 (TID 5)
20/02/14 17:58:38 INFO Executor: Running task 0.0 in stage 1.0 (TID 4)
20/02/14 17:58:38 INFO Executor: Running task 3.0 in stage 1.0 (TID 7)
20/02/14 17:58:38 INFO CodeGenerator: Code generated in 41.179 ms
20/02/14 17:58:39 INFO FileScanRDD: Reading File path: file:///edw/Reviews/SourceCode/data/AmzReviews.json, range: 66033936-83850945, partition values: [empty row]
20/02/14 17:58:39 INFO FileScanRDD: Reading File path: file:///edw/Reviews/SourceCode/data/AmzReviews.json, range: 44022624-66033936, partition values: [empty row]
20/02/14 17:58:39 INFO FileScanRDD: Reading File path: file:///edw/Reviews/SourceCode/data/AmzReviews.json, range: 22011312-44022624, partition values: [empty row]
20/02/14 17:58:39 INFO FileScanRDD: Reading File path: file:///edw/Reviews/SourceCode/data/AmzReviews.json, range: 0-22011312, partition values: [empty row]
20/02/14 17:58:39 INFO CodeGenerator: Code generated in 20.0315 ms
20/02/14 17:58:39 ERROR Executor: Exception in task 1.0 in stage 1.0 (TID 5)
java.sql.BatchUpdateException: Batch entry 0 INSERT INTO REVIEWS_STAGING ("asin","reviewrating","reviewerid","reviewername","reviewdate") VALUES ('0006487777',4.0,'AXTVARYQ2UKTA','Big D','2009-09-08 +02'::date) was aborted: ERROR: duplicate key value violates unique constraint "reviews_staging_pkey"
  Detail: Key (reviewerid, asin, reviewdate)=(AXTVARYQ2UKTA, 0006487777, 2009-09-08) already exists.  Call getNextException to see other errors in the batch.
	at org.postgresql.jdbc.BatchResultHandler.handleError(BatchResultHandler.java:153)
	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2246)
	at org.postgresql.core.v3.QueryExecutorImpl.flushIfDeadlockRisk(QueryExecutorImpl.java:1364)
	at org.postgresql.core.v3.QueryExecutorImpl.sendQuery(QueryExecutorImpl.java:1389)
	at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:495)
	at org.postgresql.jdbc.PgStatement.internalExecuteBatch(PgStatement.java:850)
	at org.postgresql.jdbc.PgStatement.executeBatch(PgStatement.java:873)
	at org.postgresql.jdbc.PgPreparedStatement.executeBatch(PgPreparedStatement.java:1569)
	at org.apache.spark.sql.execution.datasources.jdbc.JdbcUtils$.savePartition(JdbcUtils.scala:671)
	at org.apache.spark.sql.execution.datasources.jdbc.JdbcUtils$$anonfun$saveTable$1.apply(JdbcUtils.scala:838)
	at org.apache.spark.sql.execution.datasources.jdbc.JdbcUtils$$anonfun$saveTable$1.apply(JdbcUtils.scala:838)
	at org.apache.spark.rdd.RDD$$anonfun$foreachPartition$1$$anonfun$apply$28.apply(RDD.scala:980)
	at org.apache.spark.rdd.RDD$$anonfun$foreachPartition$1$$anonfun$apply$28.apply(RDD.scala:980)
	at org.apache.spark.SparkContext$$anonfun$runJob$5.apply(SparkContext.scala:2101)
	at org.apache.spark.SparkContext$$anonfun$runJob$5.apply(SparkContext.scala:2101)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)
	at org.apache.spark.scheduler.Task.run(Task.scala:123)
	at org.apache.spark.executor.Executor$TaskRunner$$anonfun$10.apply(Executor.scala:408)
	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1360)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:414)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: org.postgresql.util.PSQLException: ERROR: duplicate key value violates unique constraint "reviews_staging_pkey"
  Detail: Key (reviewerid, asin, reviewdate)=(AXTVARYQ2UKTA, 0006487777, 2009-09-08) already exists.
	at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2510)
	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2245)
	... 21 more
20/02/14 17:58:39 ERROR Executor: Exception in task 0.0 in stage 1.0 (TID 4)
java.sql.BatchUpdateException: Batch entry 0 INSERT INTO REVIEWS_STAGING ("asin","reviewrating","reviewerid","reviewername","reviewdate") VALUES ('0000000078',5.0,'A3AF8FFZAZYNE5',NULL,'2004-08-11 +02'::date) was aborted: ERROR: duplicate key value violates unique constraint "reviews_staging_pkey"
  Detail: Key (reviewerid, asin, reviewdate)=(A3AF8FFZAZYNE5, 0000000078, 2004-08-11) already exists.  Call getNextException to see other errors in the batch.
	at org.postgresql.jdbc.BatchResultHandler.handleError(BatchResultHandler.java:153)
	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2246)
	at org.postgresql.core.v3.QueryExecutorImpl.flushIfDeadlockRisk(QueryExecutorImpl.java:1364)
	at org.postgresql.core.v3.QueryExecutorImpl.sendQuery(QueryExecutorImpl.java:1389)
	at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:495)
	at org.postgresql.jdbc.PgStatement.internalExecuteBatch(PgStatement.java:850)
	at org.postgresql.jdbc.PgStatement.executeBatch(PgStatement.java:873)
	at org.postgresql.jdbc.PgPreparedStatement.executeBatch(PgPreparedStatement.java:1569)
	at org.apache.spark.sql.execution.datasources.jdbc.JdbcUtils$.savePartition(JdbcUtils.scala:671)
	at org.apache.spark.sql.execution.datasources.jdbc.JdbcUtils$$anonfun$saveTable$1.apply(JdbcUtils.scala:838)
	at org.apache.spark.sql.execution.datasources.jdbc.JdbcUtils$$anonfun$saveTable$1.apply(JdbcUtils.scala:838)
	at org.apache.spark.rdd.RDD$$anonfun$foreachPartition$1$$anonfun$apply$28.apply(RDD.scala:980)
	at org.apache.spark.rdd.RDD$$anonfun$foreachPartition$1$$anonfun$apply$28.apply(RDD.scala:980)
	at org.apache.spark.SparkContext$$anonfun$runJob$5.apply(SparkContext.scala:2101)
	at org.apache.spark.SparkContext$$anonfun$runJob$5.apply(SparkContext.scala:2101)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)
	at org.apache.spark.scheduler.Task.run(Task.scala:123)
	at org.apache.spark.executor.Executor$TaskRunner$$anonfun$10.apply(Executor.scala:408)
	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1360)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:414)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: org.postgresql.util.PSQLException: ERROR: duplicate key value violates unique constraint "reviews_staging_pkey"
  Detail: Key (reviewerid, asin, reviewdate)=(A3AF8FFZAZYNE5, 0000000078, 2004-08-11) already exists.
	at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2510)
	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2245)
	... 21 more
20/02/14 17:58:39 ERROR Executor: Exception in task 2.0 in stage 1.0 (TID 6)
java.sql.BatchUpdateException: Batch entry 0 INSERT INTO REVIEWS_STAGING ("asin","reviewrating","reviewerid","reviewername","reviewdate") VALUES ('0007172826',4.0,'A2F4L44P7ZCSOB','Diane "dianemax"','2002-07-03 +02'::date) was aborted: ERROR: duplicate key value violates unique constraint "reviews_staging_pkey"
  Detail: Key (reviewerid, asin, reviewdate)=(A2F4L44P7ZCSOB, 0007172826, 2002-07-03) already exists.  Call getNextException to see other errors in the batch.
	at org.postgresql.jdbc.BatchResultHandler.handleError(BatchResultHandler.java:153)
	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2246)
	at org.postgresql.core.v3.QueryExecutorImpl.flushIfDeadlockRisk(QueryExecutorImpl.java:1364)
	at org.postgresql.core.v3.QueryExecutorImpl.sendQuery(QueryExecutorImpl.java:1389)
	at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:495)
	at org.postgresql.jdbc.PgStatement.internalExecuteBatch(PgStatement.java:850)
	at org.postgresql.jdbc.PgStatement.executeBatch(PgStatement.java:873)
	at org.postgresql.jdbc.PgPreparedStatement.executeBatch(PgPreparedStatement.java:1569)
	at org.apache.spark.sql.execution.datasources.jdbc.JdbcUtils$.savePartition(JdbcUtils.scala:671)
	at org.apache.spark.sql.execution.datasources.jdbc.JdbcUtils$$anonfun$saveTable$1.apply(JdbcUtils.scala:838)
	at org.apache.spark.sql.execution.datasources.jdbc.JdbcUtils$$anonfun$saveTable$1.apply(JdbcUtils.scala:838)
	at org.apache.spark.rdd.RDD$$anonfun$foreachPartition$1$$anonfun$apply$28.apply(RDD.scala:980)
	at org.apache.spark.rdd.RDD$$anonfun$foreachPartition$1$$anonfun$apply$28.apply(RDD.scala:980)
	at org.apache.spark.SparkContext$$anonfun$runJob$5.apply(SparkContext.scala:2101)
	at org.apache.spark.SparkContext$$anonfun$runJob$5.apply(SparkContext.scala:2101)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)
	at org.apache.spark.scheduler.Task.run(Task.scala:123)
	at org.apache.spark.executor.Executor$TaskRunner$$anonfun$10.apply(Executor.scala:408)
	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1360)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:414)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: org.postgresql.util.PSQLException: ERROR: duplicate key value violates unique constraint "reviews_staging_pkey"
  Detail: Key (reviewerid, asin, reviewdate)=(A2F4L44P7ZCSOB, 0007172826, 2002-07-03) already exists.
	at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2510)
	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2245)
	... 21 more
20/02/14 17:58:39 ERROR Executor: Exception in task 3.0 in stage 1.0 (TID 7)
java.sql.BatchUpdateException: Batch entry 0 INSERT INTO REVIEWS_STAGING ("asin","reviewrating","reviewerid","reviewername","reviewdate") VALUES ('0007386648',5.0,'A32K9WWK36JO28','Edward G. Boyce','2011-05-12 +02'::date) was aborted: ERROR: duplicate key value violates unique constraint "reviews_staging_pkey"
  Detail: Key (reviewerid, asin, reviewdate)=(A32K9WWK36JO28, 0007386648, 2011-05-12) already exists.  Call getNextException to see other errors in the batch.
	at org.postgresql.jdbc.BatchResultHandler.handleError(BatchResultHandler.java:153)
	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2246)
	at org.postgresql.core.v3.QueryExecutorImpl.flushIfDeadlockRisk(QueryExecutorImpl.java:1364)
	at org.postgresql.core.v3.QueryExecutorImpl.sendQuery(QueryExecutorImpl.java:1389)
	at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:495)
	at org.postgresql.jdbc.PgStatement.internalExecuteBatch(PgStatement.java:850)
	at org.postgresql.jdbc.PgStatement.executeBatch(PgStatement.java:873)
	at org.postgresql.jdbc.PgPreparedStatement.executeBatch(PgPreparedStatement.java:1569)
	at org.apache.spark.sql.execution.datasources.jdbc.JdbcUtils$.savePartition(JdbcUtils.scala:671)
	at org.apache.spark.sql.execution.datasources.jdbc.JdbcUtils$$anonfun$saveTable$1.apply(JdbcUtils.scala:838)
	at org.apache.spark.sql.execution.datasources.jdbc.JdbcUtils$$anonfun$saveTable$1.apply(JdbcUtils.scala:838)
	at org.apache.spark.rdd.RDD$$anonfun$foreachPartition$1$$anonfun$apply$28.apply(RDD.scala:980)
	at org.apache.spark.rdd.RDD$$anonfun$foreachPartition$1$$anonfun$apply$28.apply(RDD.scala:980)
	at org.apache.spark.SparkContext$$anonfun$runJob$5.apply(SparkContext.scala:2101)
	at org.apache.spark.SparkContext$$anonfun$runJob$5.apply(SparkContext.scala:2101)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)
	at org.apache.spark.scheduler.Task.run(Task.scala:123)
	at org.apache.spark.executor.Executor$TaskRunner$$anonfun$10.apply(Executor.scala:408)
	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1360)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:414)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: org.postgresql.util.PSQLException: ERROR: duplicate key value violates unique constraint "reviews_staging_pkey"
  Detail: Key (reviewerid, asin, reviewdate)=(A32K9WWK36JO28, 0007386648, 2011-05-12) already exists.
	at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2510)
	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2245)
	... 21 more
20/02/14 17:58:39 WARN TaskSetManager: Lost task 1.0 in stage 1.0 (TID 5, localhost, executor driver): java.sql.BatchUpdateException: Batch entry 0 INSERT INTO REVIEWS_STAGING ("asin","reviewrating","reviewerid","reviewername","reviewdate") VALUES ('0006487777',4.0,'AXTVARYQ2UKTA','Big D','2009-09-08 +02'::date) was aborted: ERROR: duplicate key value violates unique constraint "reviews_staging_pkey"
  Detail: Key (reviewerid, asin, reviewdate)=(AXTVARYQ2UKTA, 0006487777, 2009-09-08) already exists.  Call getNextException to see other errors in the batch.
	at org.postgresql.jdbc.BatchResultHandler.handleError(BatchResultHandler.java:153)
	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2246)
	at org.postgresql.core.v3.QueryExecutorImpl.flushIfDeadlockRisk(QueryExecutorImpl.java:1364)
	at org.postgresql.core.v3.QueryExecutorImpl.sendQuery(QueryExecutorImpl.java:1389)
	at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:495)
	at org.postgresql.jdbc.PgStatement.internalExecuteBatch(PgStatement.java:850)
	at org.postgresql.jdbc.PgStatement.executeBatch(PgStatement.java:873)
	at org.postgresql.jdbc.PgPreparedStatement.executeBatch(PgPreparedStatement.java:1569)
	at org.apache.spark.sql.execution.datasources.jdbc.JdbcUtils$.savePartition(JdbcUtils.scala:671)
	at org.apache.spark.sql.execution.datasources.jdbc.JdbcUtils$$anonfun$saveTable$1.apply(JdbcUtils.scala:838)
	at org.apache.spark.sql.execution.datasources.jdbc.JdbcUtils$$anonfun$saveTable$1.apply(JdbcUtils.scala:838)
	at org.apache.spark.rdd.RDD$$anonfun$foreachPartition$1$$anonfun$apply$28.apply(RDD.scala:980)
	at org.apache.spark.rdd.RDD$$anonfun$foreachPartition$1$$anonfun$apply$28.apply(RDD.scala:980)
	at org.apache.spark.SparkContext$$anonfun$runJob$5.apply(SparkContext.scala:2101)
	at org.apache.spark.SparkContext$$anonfun$runJob$5.apply(SparkContext.scala:2101)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)
	at org.apache.spark.scheduler.Task.run(Task.scala:123)
	at org.apache.spark.executor.Executor$TaskRunner$$anonfun$10.apply(Executor.scala:408)
	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1360)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:414)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: org.postgresql.util.PSQLException: ERROR: duplicate key value violates unique constraint "reviews_staging_pkey"
  Detail: Key (reviewerid, asin, reviewdate)=(AXTVARYQ2UKTA, 0006487777, 2009-09-08) already exists.
	at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2510)
	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2245)
	... 21 more

20/02/14 17:58:39 ERROR TaskSetManager: Task 1 in stage 1.0 failed 1 times; aborting job
20/02/14 17:58:39 INFO TaskSchedulerImpl: Removed TaskSet 1.0, whose tasks have all completed, from pool 
20/02/14 17:58:39 WARN TaskSetManager: Lost task 3.0 in stage 1.0 (TID 7, localhost, executor driver): java.sql.BatchUpdateException: Batch entry 0 INSERT INTO REVIEWS_STAGING ("asin","reviewrating","reviewerid","reviewername","reviewdate") VALUES ('0007386648',5.0,'A32K9WWK36JO28','Edward G. Boyce','2011-05-12 +02'::date) was aborted: ERROR: duplicate key value violates unique constraint "reviews_staging_pkey"
  Detail: Key (reviewerid, asin, reviewdate)=(A32K9WWK36JO28, 0007386648, 2011-05-12) already exists.  Call getNextException to see other errors in the batch.
	at org.postgresql.jdbc.BatchResultHandler.handleError(BatchResultHandler.java:153)
	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2246)
	at org.postgresql.core.v3.QueryExecutorImpl.flushIfDeadlockRisk(QueryExecutorImpl.java:1364)
	at org.postgresql.core.v3.QueryExecutorImpl.sendQuery(QueryExecutorImpl.java:1389)
	at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:495)
	at org.postgresql.jdbc.PgStatement.internalExecuteBatch(PgStatement.java:850)
	at org.postgresql.jdbc.PgStatement.executeBatch(PgStatement.java:873)
	at org.postgresql.jdbc.PgPreparedStatement.executeBatch(PgPreparedStatement.java:1569)
	at org.apache.spark.sql.execution.datasources.jdbc.JdbcUtils$.savePartition(JdbcUtils.scala:671)
	at org.apache.spark.sql.execution.datasources.jdbc.JdbcUtils$$anonfun$saveTable$1.apply(JdbcUtils.scala:838)
	at org.apache.spark.sql.execution.datasources.jdbc.JdbcUtils$$anonfun$saveTable$1.apply(JdbcUtils.scala:838)
	at org.apache.spark.rdd.RDD$$anonfun$foreachPartition$1$$anonfun$apply$28.apply(RDD.scala:980)
	at org.apache.spark.rdd.RDD$$anonfun$foreachPartition$1$$anonfun$apply$28.apply(RDD.scala:980)
	at org.apache.spark.SparkContext$$anonfun$runJob$5.apply(SparkContext.scala:2101)
	at org.apache.spark.SparkContext$$anonfun$runJob$5.apply(SparkContext.scala:2101)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)
	at org.apache.spark.scheduler.Task.run(Task.scala:123)
	at org.apache.spark.executor.Executor$TaskRunner$$anonfun$10.apply(Executor.scala:408)
	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1360)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:414)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: org.postgresql.util.PSQLException: ERROR: duplicate key value violates unique constraint "reviews_staging_pkey"
  Detail: Key (reviewerid, asin, reviewdate)=(A32K9WWK36JO28, 0007386648, 2011-05-12) already exists.
	at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2510)
	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2245)
	... 21 more

20/02/14 17:58:39 INFO TaskSchedulerImpl: Removed TaskSet 1.0, whose tasks have all completed, from pool 
20/02/14 17:58:39 WARN TaskSetManager: Lost task 0.0 in stage 1.0 (TID 4, localhost, executor driver): java.sql.BatchUpdateException: Batch entry 0 INSERT INTO REVIEWS_STAGING ("asin","reviewrating","reviewerid","reviewername","reviewdate") VALUES ('0000000078',5.0,'A3AF8FFZAZYNE5',NULL,'2004-08-11 +02'::date) was aborted: ERROR: duplicate key value violates unique constraint "reviews_staging_pkey"
  Detail: Key (reviewerid, asin, reviewdate)=(A3AF8FFZAZYNE5, 0000000078, 2004-08-11) already exists.  Call getNextException to see other errors in the batch.
	at org.postgresql.jdbc.BatchResultHandler.handleError(BatchResultHandler.java:153)
	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2246)
	at org.postgresql.core.v3.QueryExecutorImpl.flushIfDeadlockRisk(QueryExecutorImpl.java:1364)
	at org.postgresql.core.v3.QueryExecutorImpl.sendQuery(QueryExecutorImpl.java:1389)
	at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:495)
	at org.postgresql.jdbc.PgStatement.internalExecuteBatch(PgStatement.java:850)
	at org.postgresql.jdbc.PgStatement.executeBatch(PgStatement.java:873)
	at org.postgresql.jdbc.PgPreparedStatement.executeBatch(PgPreparedStatement.java:1569)
	at org.apache.spark.sql.execution.datasources.jdbc.JdbcUtils$.savePartition(JdbcUtils.scala:671)
	at org.apache.spark.sql.execution.datasources.jdbc.JdbcUtils$$anonfun$saveTable$1.apply(JdbcUtils.scala:838)
	at org.apache.spark.sql.execution.datasources.jdbc.JdbcUtils$$anonfun$saveTable$1.apply(JdbcUtils.scala:838)
	at org.apache.spark.rdd.RDD$$anonfun$foreachPartition$1$$anonfun$apply$28.apply(RDD.scala:980)
	at org.apache.spark.rdd.RDD$$anonfun$foreachPartition$1$$anonfun$apply$28.apply(RDD.scala:980)
	at org.apache.spark.SparkContext$$anonfun$runJob$5.apply(SparkContext.scala:2101)
	at org.apache.spark.SparkContext$$anonfun$runJob$5.apply(SparkContext.scala:2101)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)
	at org.apache.spark.scheduler.Task.run(Task.scala:123)
	at org.apache.spark.executor.Executor$TaskRunner$$anonfun$10.apply(Executor.scala:408)
	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1360)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:414)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: org.postgresql.util.PSQLException: ERROR: duplicate key value violates unique constraint "reviews_staging_pkey"
  Detail: Key (reviewerid, asin, reviewdate)=(A3AF8FFZAZYNE5, 0000000078, 2004-08-11) already exists.
	at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2510)
	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2245)
	... 21 more

20/02/14 17:58:39 INFO TaskSchedulerImpl: Removed TaskSet 1.0, whose tasks have all completed, from pool 
20/02/14 17:58:39 WARN TaskSetManager: Lost task 2.0 in stage 1.0 (TID 6, localhost, executor driver): java.sql.BatchUpdateException: Batch entry 0 INSERT INTO REVIEWS_STAGING ("asin","reviewrating","reviewerid","reviewername","reviewdate") VALUES ('0007172826',4.0,'A2F4L44P7ZCSOB','Diane "dianemax"','2002-07-03 +02'::date) was aborted: ERROR: duplicate key value violates unique constraint "reviews_staging_pkey"
  Detail: Key (reviewerid, asin, reviewdate)=(A2F4L44P7ZCSOB, 0007172826, 2002-07-03) already exists.  Call getNextException to see other errors in the batch.
	at org.postgresql.jdbc.BatchResultHandler.handleError(BatchResultHandler.java:153)
	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2246)
	at org.postgresql.core.v3.QueryExecutorImpl.flushIfDeadlockRisk(QueryExecutorImpl.java:1364)
	at org.postgresql.core.v3.QueryExecutorImpl.sendQuery(QueryExecutorImpl.java:1389)
	at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:495)
	at org.postgresql.jdbc.PgStatement.internalExecuteBatch(PgStatement.java:850)
	at org.postgresql.jdbc.PgStatement.executeBatch(PgStatement.java:873)
	at org.postgresql.jdbc.PgPreparedStatement.executeBatch(PgPreparedStatement.java:1569)
	at org.apache.spark.sql.execution.datasources.jdbc.JdbcUtils$.savePartition(JdbcUtils.scala:671)
	at org.apache.spark.sql.execution.datasources.jdbc.JdbcUtils$$anonfun$saveTable$1.apply(JdbcUtils.scala:838)
	at org.apache.spark.sql.execution.datasources.jdbc.JdbcUtils$$anonfun$saveTable$1.apply(JdbcUtils.scala:838)
	at org.apache.spark.rdd.RDD$$anonfun$foreachPartition$1$$anonfun$apply$28.apply(RDD.scala:980)
	at org.apache.spark.rdd.RDD$$anonfun$foreachPartition$1$$anonfun$apply$28.apply(RDD.scala:980)
	at org.apache.spark.SparkContext$$anonfun$runJob$5.apply(SparkContext.scala:2101)
	at org.apache.spark.SparkContext$$anonfun$runJob$5.apply(SparkContext.scala:2101)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)
	at org.apache.spark.scheduler.Task.run(Task.scala:123)
	at org.apache.spark.executor.Executor$TaskRunner$$anonfun$10.apply(Executor.scala:408)
	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1360)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:414)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: org.postgresql.util.PSQLException: ERROR: duplicate key value violates unique constraint "reviews_staging_pkey"
  Detail: Key (reviewerid, asin, reviewdate)=(A2F4L44P7ZCSOB, 0007172826, 2002-07-03) already exists.
	at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2510)
	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2245)
	... 21 more

20/02/14 17:58:39 INFO TaskSchedulerImpl: Removed TaskSet 1.0, whose tasks have all completed, from pool 
20/02/14 17:58:39 INFO TaskSchedulerImpl: Cancelling stage 1
20/02/14 17:58:39 INFO TaskSchedulerImpl: Killing all running tasks in stage 1: Stage cancelled
20/02/14 17:58:39 INFO DAGScheduler: ResultStage 1 (jdbc at NativeMethodAccessorImpl.java:0) failed in 1.027 s due to Job aborted due to stage failure: Task 1 in stage 1.0 failed 1 times, most recent failure: Lost task 1.0 in stage 1.0 (TID 5, localhost, executor driver): java.sql.BatchUpdateException: Batch entry 0 INSERT INTO REVIEWS_STAGING ("asin","reviewrating","reviewerid","reviewername","reviewdate") VALUES ('0006487777',4.0,'AXTVARYQ2UKTA','Big D','2009-09-08 +02'::date) was aborted: ERROR: duplicate key value violates unique constraint "reviews_staging_pkey"
  Detail: Key (reviewerid, asin, reviewdate)=(AXTVARYQ2UKTA, 0006487777, 2009-09-08) already exists.  Call getNextException to see other errors in the batch.
	at org.postgresql.jdbc.BatchResultHandler.handleError(BatchResultHandler.java:153)
	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2246)
	at org.postgresql.core.v3.QueryExecutorImpl.flushIfDeadlockRisk(QueryExecutorImpl.java:1364)
	at org.postgresql.core.v3.QueryExecutorImpl.sendQuery(QueryExecutorImpl.java:1389)
	at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:495)
	at org.postgresql.jdbc.PgStatement.internalExecuteBatch(PgStatement.java:850)
	at org.postgresql.jdbc.PgStatement.executeBatch(PgStatement.java:873)
	at org.postgresql.jdbc.PgPreparedStatement.executeBatch(PgPreparedStatement.java:1569)
	at org.apache.spark.sql.execution.datasources.jdbc.JdbcUtils$.savePartition(JdbcUtils.scala:671)
	at org.apache.spark.sql.execution.datasources.jdbc.JdbcUtils$$anonfun$saveTable$1.apply(JdbcUtils.scala:838)
	at org.apache.spark.sql.execution.datasources.jdbc.JdbcUtils$$anonfun$saveTable$1.apply(JdbcUtils.scala:838)
	at org.apache.spark.rdd.RDD$$anonfun$foreachPartition$1$$anonfun$apply$28.apply(RDD.scala:980)
	at org.apache.spark.rdd.RDD$$anonfun$foreachPartition$1$$anonfun$apply$28.apply(RDD.scala:980)
	at org.apache.spark.SparkContext$$anonfun$runJob$5.apply(SparkContext.scala:2101)
	at org.apache.spark.SparkContext$$anonfun$runJob$5.apply(SparkContext.scala:2101)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)
	at org.apache.spark.scheduler.Task.run(Task.scala:123)
	at org.apache.spark.executor.Executor$TaskRunner$$anonfun$10.apply(Executor.scala:408)
	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1360)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:414)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: org.postgresql.util.PSQLException: ERROR: duplicate key value violates unique constraint "reviews_staging_pkey"
  Detail: Key (reviewerid, asin, reviewdate)=(AXTVARYQ2UKTA, 0006487777, 2009-09-08) already exists.
	at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2510)
	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2245)
	... 21 more

Driver stacktrace:
20/02/14 17:58:39 INFO DAGScheduler: Job 1 failed: jdbc at NativeMethodAccessorImpl.java:0, took 1.034818 s
20/02/14 17:58:39 INFO BlockManagerInfo: Removed broadcast_3_piece0 on 192.168.1.25:3091 in memory (size: 29.4 KB, free: 366.2 MB)
20/02/14 17:58:39 INFO BlockManagerInfo: Removed broadcast_2_piece0 on 192.168.1.25:3091 in memory (size: 23.3 KB, free: 366.3 MB)
20/02/14 17:58:39 INFO ContextCleaner: Cleaned accumulator 35
20/02/14 17:58:39 INFO ContextCleaner: Cleaned accumulator 34
20/02/14 17:58:39 INFO ContextCleaner: Cleaned accumulator 32
20/02/14 17:58:39 INFO ContextCleaner: Cleaned accumulator 31
20/02/14 17:58:39 INFO ContextCleaner: Cleaned accumulator 33
Traceback (most recent call last):
  File "/edw/Reviews/SourceCode/python/rating_ods_load.py", line 31, in <module>
    Cl_Reviews_df.write.jdbc(url=url, table="REVIEWS_STAGING", mode=mode, properties=properties)
  File "/edw/Reviews/spark-2.4.5-bin-hadoop2.7/python/lib/pyspark.zip/pyspark/sql/readwriter.py", line 987, in jdbc
  File "/edw/Reviews/spark-2.4.5-bin-hadoop2.7/python/lib/py4j-0.10.7-src.zip/py4j/java_gateway.py", line 1257, in __call__
  File "/edw/Reviews/spark-2.4.5-bin-hadoop2.7/python/lib/pyspark.zip/pyspark/sql/utils.py", line 63, in deco
  File "/edw/Reviews/spark-2.4.5-bin-hadoop2.7/python/lib/py4j-0.10.7-src.zip/py4j/protocol.py", line 328, in get_return_value
py4j.protocol.Py4JJavaError: An error occurred while calling o52.jdbc.
: org.apache.spark.SparkException: Job aborted due to stage failure: Task 1 in stage 1.0 failed 1 times, most recent failure: Lost task 1.0 in stage 1.0 (TID 5, localhost, executor driver): java.sql.BatchUpdateException: Batch entry 0 INSERT INTO REVIEWS_STAGING ("asin","reviewrating","reviewerid","reviewername","reviewdate") VALUES ('0006487777',4.0,'AXTVARYQ2UKTA','Big D','2009-09-08 +02'::date) was aborted: ERROR: duplicate key value violates unique constraint "reviews_staging_pkey"
  Detail: Key (reviewerid, asin, reviewdate)=(AXTVARYQ2UKTA, 0006487777, 2009-09-08) already exists.  Call getNextException to see other errors in the batch.
	at org.postgresql.jdbc.BatchResultHandler.handleError(BatchResultHandler.java:153)
	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2246)
	at org.postgresql.core.v3.QueryExecutorImpl.flushIfDeadlockRisk(QueryExecutorImpl.java:1364)
	at org.postgresql.core.v3.QueryExecutorImpl.sendQuery(QueryExecutorImpl.java:1389)
	at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:495)
	at org.postgresql.jdbc.PgStatement.internalExecuteBatch(PgStatement.java:850)
	at org.postgresql.jdbc.PgStatement.executeBatch(PgStatement.java:873)
	at org.postgresql.jdbc.PgPreparedStatement.executeBatch(PgPreparedStatement.java:1569)
	at org.apache.spark.sql.execution.datasources.jdbc.JdbcUtils$.savePartition(JdbcUtils.scala:671)
	at org.apache.spark.sql.execution.datasources.jdbc.JdbcUtils$$anonfun$saveTable$1.apply(JdbcUtils.scala:838)
	at org.apache.spark.sql.execution.datasources.jdbc.JdbcUtils$$anonfun$saveTable$1.apply(JdbcUtils.scala:838)
	at org.apache.spark.rdd.RDD$$anonfun$foreachPartition$1$$anonfun$apply$28.apply(RDD.scala:980)
	at org.apache.spark.rdd.RDD$$anonfun$foreachPartition$1$$anonfun$apply$28.apply(RDD.scala:980)
	at org.apache.spark.SparkContext$$anonfun$runJob$5.apply(SparkContext.scala:2101)
	at org.apache.spark.SparkContext$$anonfun$runJob$5.apply(SparkContext.scala:2101)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)
	at org.apache.spark.scheduler.Task.run(Task.scala:123)
	at org.apache.spark.executor.Executor$TaskRunner$$anonfun$10.apply(Executor.scala:408)
	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1360)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:414)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: org.postgresql.util.PSQLException: ERROR: duplicate key value violates unique constraint "reviews_staging_pkey"
  Detail: Key (reviewerid, asin, reviewdate)=(AXTVARYQ2UKTA, 0006487777, 2009-09-08) already exists.
	at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2510)
	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2245)
	... 21 more

Driver stacktrace:
	at org.apache.spark.scheduler.DAGScheduler.org$apache$spark$scheduler$DAGScheduler$$failJobAndIndependentStages(DAGScheduler.scala:1891)
	at org.apache.spark.scheduler.DAGScheduler$$anonfun$abortStage$1.apply(DAGScheduler.scala:1879)
	at org.apache.spark.scheduler.DAGScheduler$$anonfun$abortStage$1.apply(DAGScheduler.scala:1878)
	at scala.collection.mutable.ResizableArray$class.foreach(ResizableArray.scala:59)
	at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:48)
	at org.apache.spark.scheduler.DAGScheduler.abortStage(DAGScheduler.scala:1878)
	at org.apache.spark.scheduler.DAGScheduler$$anonfun$handleTaskSetFailed$1.apply(DAGScheduler.scala:927)
	at org.apache.spark.scheduler.DAGScheduler$$anonfun$handleTaskSetFailed$1.apply(DAGScheduler.scala:927)
	at scala.Option.foreach(Option.scala:257)
	at org.apache.spark.scheduler.DAGScheduler.handleTaskSetFailed(DAGScheduler.scala:927)
	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.doOnReceive(DAGScheduler.scala:2112)
	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:2061)
	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:2050)
	at org.apache.spark.util.EventLoop$$anon$1.run(EventLoop.scala:49)
	at org.apache.spark.scheduler.DAGScheduler.runJob(DAGScheduler.scala:738)
	at org.apache.spark.SparkContext.runJob(SparkContext.scala:2061)
	at org.apache.spark.SparkContext.runJob(SparkContext.scala:2082)
	at org.apache.spark.SparkContext.runJob(SparkContext.scala:2101)
	at org.apache.spark.SparkContext.runJob(SparkContext.scala:2126)
	at org.apache.spark.rdd.RDD$$anonfun$foreachPartition$1.apply(RDD.scala:980)
	at org.apache.spark.rdd.RDD$$anonfun$foreachPartition$1.apply(RDD.scala:978)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:112)
	at org.apache.spark.rdd.RDD.withScope(RDD.scala:385)
	at org.apache.spark.rdd.RDD.foreachPartition(RDD.scala:978)
	at org.apache.spark.sql.execution.datasources.jdbc.JdbcUtils$.saveTable(JdbcUtils.scala:838)
	at org.apache.spark.sql.execution.datasources.jdbc.JdbcRelationProvider.createRelation(JdbcRelationProvider.scala:68)
	at org.apache.spark.sql.execution.datasources.SaveIntoDataSourceCommand.run(SaveIntoDataSourceCommand.scala:45)
	at org.apache.spark.sql.execution.command.ExecutedCommandExec.sideEffectResult$lzycompute(commands.scala:70)
	at org.apache.spark.sql.execution.command.ExecutedCommandExec.sideEffectResult(commands.scala:68)
	at org.apache.spark.sql.execution.command.ExecutedCommandExec.doExecute(commands.scala:86)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$execute$1.apply(SparkPlan.scala:131)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$execute$1.apply(SparkPlan.scala:127)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$executeQuery$1.apply(SparkPlan.scala:155)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
	at org.apache.spark.sql.execution.SparkPlan.executeQuery(SparkPlan.scala:152)
	at org.apache.spark.sql.execution.SparkPlan.execute(SparkPlan.scala:127)
	at org.apache.spark.sql.execution.QueryExecution.toRdd$lzycompute(QueryExecution.scala:83)
	at org.apache.spark.sql.execution.QueryExecution.toRdd(QueryExecution.scala:81)
	at org.apache.spark.sql.DataFrameWriter$$anonfun$runCommand$1.apply(DataFrameWriter.scala:676)
	at org.apache.spark.sql.DataFrameWriter$$anonfun$runCommand$1.apply(DataFrameWriter.scala:676)
	at org.apache.spark.sql.execution.SQLExecution$$anonfun$withNewExecutionId$1.apply(SQLExecution.scala:80)
	at org.apache.spark.sql.execution.SQLExecution$.withSQLConfPropagated(SQLExecution.scala:127)
	at org.apache.spark.sql.execution.SQLExecution$.withNewExecutionId(SQLExecution.scala:75)
	at org.apache.spark.sql.DataFrameWriter.runCommand(DataFrameWriter.scala:676)
	at org.apache.spark.sql.DataFrameWriter.saveToV1Source(DataFrameWriter.scala:285)
	at org.apache.spark.sql.DataFrameWriter.save(DataFrameWriter.scala:271)
	at org.apache.spark.sql.DataFrameWriter.jdbc(DataFrameWriter.scala:515)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:244)
	at py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:357)
	at py4j.Gateway.invoke(Gateway.java:282)
	at py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:132)
	at py4j.commands.CallCommand.execute(CallCommand.java:79)
	at py4j.GatewayConnection.run(GatewayConnection.java:238)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.sql.BatchUpdateException: Batch entry 0 INSERT INTO REVIEWS_STAGING ("asin","reviewrating","reviewerid","reviewername","reviewdate") VALUES ('0006487777',4.0,'AXTVARYQ2UKTA','Big D','2009-09-08 +02'::date) was aborted: ERROR: duplicate key value violates unique constraint "reviews_staging_pkey"
  Detail: Key (reviewerid, asin, reviewdate)=(AXTVARYQ2UKTA, 0006487777, 2009-09-08) already exists.  Call getNextException to see other errors in the batch.
	at org.postgresql.jdbc.BatchResultHandler.handleError(BatchResultHandler.java:153)
	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2246)
	at org.postgresql.core.v3.QueryExecutorImpl.flushIfDeadlockRisk(QueryExecutorImpl.java:1364)
	at org.postgresql.core.v3.QueryExecutorImpl.sendQuery(QueryExecutorImpl.java:1389)
	at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:495)
	at org.postgresql.jdbc.PgStatement.internalExecuteBatch(PgStatement.java:850)
	at org.postgresql.jdbc.PgStatement.executeBatch(PgStatement.java:873)
	at org.postgresql.jdbc.PgPreparedStatement.executeBatch(PgPreparedStatement.java:1569)
	at org.apache.spark.sql.execution.datasources.jdbc.JdbcUtils$.savePartition(JdbcUtils.scala:671)
	at org.apache.spark.sql.execution.datasources.jdbc.JdbcUtils$$anonfun$saveTable$1.apply(JdbcUtils.scala:838)
	at org.apache.spark.sql.execution.datasources.jdbc.JdbcUtils$$anonfun$saveTable$1.apply(JdbcUtils.scala:838)
	at org.apache.spark.rdd.RDD$$anonfun$foreachPartition$1$$anonfun$apply$28.apply(RDD.scala:980)
	at org.apache.spark.rdd.RDD$$anonfun$foreachPartition$1$$anonfun$apply$28.apply(RDD.scala:980)
	at org.apache.spark.SparkContext$$anonfun$runJob$5.apply(SparkContext.scala:2101)
	at org.apache.spark.SparkContext$$anonfun$runJob$5.apply(SparkContext.scala:2101)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)
	at org.apache.spark.scheduler.Task.run(Task.scala:123)
	at org.apache.spark.executor.Executor$TaskRunner$$anonfun$10.apply(Executor.scala:408)
	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1360)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:414)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	... 1 more
Caused by: org.postgresql.util.PSQLException: ERROR: duplicate key value violates unique constraint "reviews_staging_pkey"
  Detail: Key (reviewerid, asin, reviewdate)=(AXTVARYQ2UKTA, 0006487777, 2009-09-08) already exists.
	at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2510)
	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2245)
	... 21 more

20/02/14 17:58:39 INFO SparkContext: Invoking stop() from shutdown hook
20/02/14 17:58:40 INFO SparkUI: Stopped Spark web UI at http://192.168.1.25:4040
20/02/14 17:58:40 INFO MapOutputTrackerMasterEndpoint: MapOutputTrackerMasterEndpoint stopped!
20/02/14 17:58:40 INFO MemoryStore: MemoryStore cleared
20/02/14 17:58:40 INFO BlockManager: BlockManager stopped
20/02/14 17:58:40 INFO BlockManagerMaster: BlockManagerMaster stopped
20/02/14 17:58:40 INFO OutputCommitCoordinator$OutputCommitCoordinatorEndpoint: OutputCommitCoordinator stopped!
20/02/14 17:58:40 INFO SparkContext: Successfully stopped SparkContext
20/02/14 17:58:40 INFO ShutdownHookManager: Shutdown hook called
20/02/14 17:58:40 INFO ShutdownHookManager: Deleting directory /tmp/spark-61c46695-2f9e-42c0-9dc8-96eb6f4fc9ea
20/02/14 17:58:40 INFO ShutdownHookManager: Deleting directory /tmp/spark-1afdd9d6-391f-44a5-b8c3-6ad7ed31d88a
20/02/14 17:58:40 INFO ShutdownHookManager: Deleting directory /tmp/spark-1afdd9d6-391f-44a5-b8c3-6ad7ed31d88a/pyspark-48db74e7-7d1d-49ba-a9fb-2e26e79bf835
